text/0000-ledger-api.md

```diff
@@ -0,0 +1,380 @@
+
+The 'Ledger' is, therefore, a primary abstraction in a blockchain system, including in Hyperledger Fabric; this abstraction allows the concepts of states, world state, collections, transactions, blocks etc. to be tied together and formalised. 
+
+Looking at the objects that make up this conceptual architecture, a full set of objects would have the following relationships. 
+
+![](0000-ledger-api-relationships.png)
+
+Note that this RFC is concerned with the ledger-api and the ledger,collection and state; the 'white' boxes will be subsequent RFCs.
+
+## Ledger
+
+A Ledger's current value is held in a set of collections. A single ledger is available to any given deployed contract, though other ledgers may be available within the Network Channel and can be accessed via other contracts
+
+Of the collections that are reference by the ledger, there a single collection in the ledger that is public (the 'WorldState') and held by every organization in the Network Channel. A ledger may have 1 or more private collections that are available only in some organizations. Starting at Fabric v2.0 there will be an implicit private collection per organization.
+
+## Collection
+
+A collection is a set of states, with each state holding a business object or data. Each state being addressed by a key. Private collections are identified by name, and held within a set of policy-defined organizations.
+
+Accessing any given collection is via the Ledger instance. 
+
+```java
+	Ledger = Ledger.getLedger(ctx)
+
+	// predefined constant for the world state
+	Collection worldState = Ledger.getCollection(Collection.WORLD);
+
+	// Private Data collections are reference by name
+	Collection privateData = Ledger.getCollection("PrivateSalesData");
+
+	// Each organization will have an implicit private data collection created for them
+	// this has a prefixed name
+	Collection org1PrivateData = Ledger.getCollection(Collection.organizationCollection("orgsmspID"));
+
+	// with a convience version of
+	Collection org1PrivateData = Ledger.getOrganizationCollection("orgsmspID");
+
+```
+
+There are underlying technical restrictions on some operations. For example you cannot combine private data range queries and updates in a transaction.
+Nor do private data collections support pagination. Should these not be supported, a 'not-supported' error/exception will be thrown.
+
+As there are two essential types of collection - public and private - this is available via an enumerated value. Each languages approach to enumerations is syntactical different but conceptually similar
```

Comment:
Some of the comment that I mentioned above is applicable here.

Reply:
It is true that the word collection is being overloaded?  Is there an alternative you can think of?

Reply:
We are trying to put the following three variants under a single bucket called a collection. I am completely fine with that provided that it does not create any confusion for the developers. 

1. **Private collections**: only a sub-set of channel members are part of this collection. It provides data privacy within a channel. 
2. **Public collections**: everyone in the channel is a member of this collection. This allows us to have different endorsement policy for public data (coarse-granular compared to a key-based endorsement but fine-granular compared to chaincode level endorsement policy). Further, it helps in avoiding the data leak at the ordering service, for example, if the deployment uses a thrid-party ordering service (provided by some cloud vendors). Note that this concept is not explicitly discussed in Fabric but we can create such a collection even now as the only difference compared to a private collection is the member list. We did discuss the concept of the public collection while designing the private data but didn't write it on any docs. In one of the comments above, Dave mentioned about bringing the public collection concept to core fabric. 
3. **World state**: States that are visible to everyone in the channel including the ordering service. This is the state that actually presents in the blocks. We can also term this as blockchain states or ledger data. This includes the hash of the private data which can be accessed by everyone in the channel through `GetPrivateDataHash()`. Further, a few patterns have been introduced by Dave using this new API which makes the developer aware of hash of some private data belonging to a sub-set of organizations is present in the blockchain.  

IMO, public and private collection can be called a collection without the actual differentiation as the underlying concept is the same (these data do not get into blocks) but there is a slight difference in the goal. Given that, we might call the world state as public collection and the other as a private collection. Again, I am not sure whether it would confuse the developers.